### PR TITLE
feat: add completion to pre-filtering

### PR DESCRIPTION
**Pull request by fdschmidt93**
This PR adds simple insert mode completion to `tag` prefiltering to both (a) cycle through (in e.g. empty prompt or prompt merely comprises a tag) or (b) incrementally complete `tags` by

1. Collecting `tags` in a set table during scoring
2. Trigger `action` to set insert completion for prompt with `complete` **

https://user-images.githubusercontent.com/39233597/110256294-57385c00-7f98-11eb-86e9-9f647bb2a659.mp4

Accepting a match currently is done via (a)`<C-y>` (standard vim binding), (b) continue typing with open menu or (c) `ESC` as `<CR>` is of course bound to jumping to line of entry. 

Known issue:

- Backspace freezes after going into normal mode (e: i.e. accepting completion with `ESC`): one can still cycle through but only delete prompt from normal mode (i.e. `cc`) which I guess is related to #261 

** Unlike `:help complete` suggests, it works without `<C-R>` as evident by mapping set to `<C-l>` :)

### CODE REVIEW

File: lua/telescope/actions/init.lua
Comment (Conni2461) [inline lua/telescope/actions/init.lua:431]: Cool feature. I will look at it closer later.
But thats one thing i immediately noticed. This will not work if the prompt is different or not existing.
There is `action_state.get_current_line()` or so that will do give you the line without `prompt_prefix`

Comment (fdschmidt93) [reply lua/telescope/actions/init.lua]: Thanks! Changed accordingly :)

File: lua/telescope/actions/init.lua
Comment (Conni2461) [inline lua/telescope/actions/init.lua:429]: I am not sure if we should not just exit silently when someone presses `<c-l>`, for example on accident in `find_files`.

File: lua/telescope/utils.lua
Comment (Conni2461) [inline lua/telescope/utils.lua:399]: Just thinking out load. There is also `__newindex` do we need to handle that as well?

**Comment by Conni2461** [comment]
> Backspace freezes after going into normal mode (e: i.e. accepting completion with ESC): one can still cycle through but only delete prompt from normal mode (i.e. cc) which I guess is related to #261

Yeah its a neovim upstream issue with prompt buffers. So not setting a `prompt_prefix` fixes that. The vim patch that addressed this issue was `v8.2.1976` but that one cant port until this was ported first:
```
[conni@home neovim]$ ./scripts/vim-patch.sh -L --grep prompt -- src
v8.1.0035
v8.1.0042
v8.1.0064
v8.1.1813
v8.2.0052
v8.2.0198
v8.2.0301
v8.2.0911
v8.2.0946
v8.2.0948
v8.2.0950
v8.2.1147
v8.2.1588
v8.2.1781
v8.2.1844
v8.2.1976 <- This patch fixes your issue
v8.2.2014
```

And the first one `v8.1.0035` is not that easy to port. See problems here: https://github.com/neovim/neovim/pull/13027

File: lua/telescope/actions/init.lua
Comment (fdschmidt93) [reply lua/telescope/actions/init.lua:429]: I initially followed your line of thought though then personally preferred verbosity plus possibly accounting for future scenarios I can't think of yet.

Nevertheless, your point of view is totally fair since, at least for the time being, there should not be scenarios in which `actions.complete_tag` can be intentionally called _and_ `tags` can be an empty table. An empty table should also eventually trigger `print(No matches found.)`.

So, let's settle for the below?

```lua
  local current_picker = action_state.get_current_picker(prompt_bufnr)
  local tags = current_picker.sorter.tags
  local delimiter = current_picker.sorter._delimiter

  -- silently exit for pickers that do not prefilter
  if not tags then return end
```

Comment (Conni2461) [reply lua/telescope/actions/init.lua:429]: Yeah i know i said hey lets give user feedback if he calls `Telescope lsp_*_diagnostics` but i am not sure how we should handle it here. Like someone hits it accidentally in `find_files` and just sees a random message about `tags`. 

But yeah i see what you saying, i also never know how to handle these cases

File: lua/telescope/utils.lua
Comment (fdschmidt93) [reply lua/telescope/utils.lua:399]: No, from reading documentation my understanding is that we don't need it here. Haven't used it and I quickly tried reeimplementing a set with `__newindex`. I might be clumsy but that use case is circular, since `__newindex` defines behavior for keys that are absent from the table. When trying to set a key for a table that's not there in `__newindex` it gets reinvoked. :sweat_smile: 

Sure thing :)

File: lua/telescope/actions/init.lua
Comment (fdschmidt93) [reply lua/telescope/actions/init.lua:429]: No diagnostics for `lsp_*_diagnostics` is a much more common case though. Let's not confuse the 99% of telescope users still unaware of prefiltering with a half-hearted message as per the current commit they lack context to understand and be explicit instead

```lua
  local current_picker = action_state.get_current_picker(prompt_bufnr)
  local tags = current_picker.sorter.tags
  local delimiter = current_picker.sorter._delimiter

  if not tags then
    print('No tag pre-filtering set for this picker.')
    return
  end
```

since empty tables on second thought rightfully go to `print('No matches found.')`

If you'd agree we could additionally point to pickers for which tag pre-filtering is set out-of-the-box as a gimmick to increase discoverability of the feature like
`print('No tag pre-filtering set for this picker. Try, for instance, <C-l> for lsp_*_diagnostics or lsp_*_symbols!)` or something similar, since accidentally hitting `<C-l>` is also not that common I suppose, but the above is fine of course, too.

Comment (Conni2461) [reply lua/telescope/actions/init.lua]: Yeah works now, thanks :)
I was also thinking maybe we should just map that action for functions that actually support pre-filtering.

I am going to test it now :)

**Review by Conni2461 — APPROVED** [review]
LGTM i like it :)

Thanks :)

I will wait for an answer to my last two comments after that i would merge

File: lua/telescope/utils.lua
Comment (fdschmidt93) [reply lua/telescope/utils.lua:395]: I fully understand - required but not sure where to put it :laughing: , let's do this local in `sorters.lua` then :)

**Pull request merged**
